lib/Sema/TypeCheckType.cpp

```diff
@@ -4291,19 +4291,22 @@ TypeResolver::resolveOwnershipTypeRepr(OwnershipTypeRepr *repr,
       options.hasBase(TypeResolverContext::SubscriptDecl) ||
       options.hasBase(TypeResolverContext::EnumElementDecl)) {
 
-    decltype(diag::attr_only_on_parameters) diagID;
-    if (options.getBaseContext() == TypeResolverContext::SubscriptDecl) {
-      diagID = diag::attr_not_on_subscript_parameters;
-    } else if (options.is(TypeResolverContext::VariadicFunctionInput)) {
-      diagID = diag::attr_not_on_variadic_parameters;
-    } else {
-      diagID = diag::attr_only_on_parameters;
-    }
     StringRef name;
     if (ownershipRepr) {
       name = ownershipRepr->getSpecifierSpelling();
     }
-    diagnoseInvalid(repr, repr->getSpecifierLoc(), diagID, name);
+
+    if (options.hasBase(TypeResolverContext::SubscriptDecl) ||
+        options.is(TypeResolverContext::VariadicFunctionInput) ||
+        options.hasBase(TypeResolverContext::EnumElementDecl)) {
+      diagnoseInvalid(repr, repr->getSpecifierLoc(),
+                      diag::attr_must_be_on_valid_parameters, name,
+                      "function or initializer");
```

Comment:
I think you can inline this into the diagnostic text instead of having to pass this as a string because it's static.